### ISSUE TITLE
Warn on shadowing built-in types

### ISSUE THREAD

**Issue by Wilfred** [rust-lang/rust#36046]
I got a little confused by the following code:

```rust
struct Foo<T> {
    value: T
}

impl<u32> Foo<u32> {
    fn impl_func(&self) -> u32 {
        42
    }
}
```

rustc gives the error: `42: expected type parameter, found integral variable`.

The mistake is that I've created a type parameter named `u32`. Could we warn on this case? I can't see any legitimate use case for shadowing `u32`.

**Comment by nagisa** [comment]
Should report an issue to [clippy](https://github.com/Manishearth/rust-clippy) in case they don’t already have such a lint instead.

**Comment by Wilfred** [comment]
Since this code is also a compile error, I believe clippy can't help.

**Comment by mcarton** [comment]
@Wilfred Sure it can. Purely syntactic lints are run before type checking.

This had already been reported on Clippy: https://github.com/Manishearth/rust-clippy/issues/949. I implemented it: https://github.com/Manishearth/rust-clippy/pull/1199.

**Comment by Wilfred** [comment]
Aha, this is well-trodden ground then. I'm going to close as a duplicate of #35030.

Huge thanks for implementing it in clippy! I did not know clippy could help with invalid code.
